### PR TITLE
Fix animated mesh properties

### PR DESCRIPTION
**Pull request by judubu**
## Purpose of this PR

**Ticket/Jira #:**

Sanitizing mesh data requires the original topology (face count and face indices) to be available at all times.
But to improve deserialization performance when reading animation data the static properties are filtered out and not available to the import logic.

A ConversionState interface has been added to hold Sample states and is stored inside the AccessMask initialized globally for all prims at the first frame of animation.

UnitTests to come

## Testing

**Functional Testing status:**

**Performance Testing status:**

## Overall Product Risks

**Complexity:**

**Halo Effect:**

## Additional information

**Note to reviewers:**

**Reminder:**

- [ ] Add entry in CHANGELOG.md _(if applicable)_

### CODE REVIEW

File: package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs [hunk 1]
```diff
@@ -55,6 +55,24 @@ public class MeshImportStrategy : IImporter
...
+            //     var hasPrimSource = primMap[meshPath].TryGetComponent(out primSource);
+            //     if (hasPrimSource)
+            //     {
+            //         contexts[i] = primSource;
+            //     }
+            //     else
+            //     {
+            //         contexts[i] = null;
+            //     }
+            // }
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs]: Delete?

**Comment by jcowles** [comment]
Seems like this is still not importing UVs correctly - is that a separate issue?

Test model:
https://sketchfab.com/3d-models/grass-variations-c84573687bf14f89938002df4ca0e696#download

3.0.0 FIX branch:
image
image


2.0.0
image
image


I hope to send review for PR this weekend

**Comment by jcowles** [comment]
Discussing offline, I think the right thing to do is to first land this change and then land the UV fix after.

File: package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs [hunk 2]
```diff
@@ -535,7 +559,8 @@ public static void ImportSkinning(string path, SanitizedMeshSample usdMesh, Mesh
                 unityMesh.normals = usdMesh.normals;
                 Profiler.EndSample(); // Import Normals
             }
-            else if (ShouldCompute(options.meshOptions.tangents))
+            // If the mesh had normals have been sanitized don't
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs]: This comment makes no sense to me :) maybe it's incomplete?

File: package/com.unity.formats.usd/Dependencies/USD.NET/serialization/AccessMask.cs [hunk 3]
```diff
@@ -12,13 +12,43 @@
...
+    /// downstream logic needs some static properties they can be stored temporarily in a ConversionContext.
+    /// </remarks>
+    public interface IConversionState
+    {
+    }
+
+    /// <summary>
+    /// This class holds the data required by the deserializer to optimize reading of primitive with dynamic properties
+    /// </summary>
+    public class DeserializationContext : IEnumerable
```
Comment (jcowles) [inline package/com.unity.formats.usd/Dependencies/USD.NET/serialization/AccessMask.cs]: Would `StreamingContext` be a more accurate name?

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs [hunk 4]
```diff
@@ -409,15 +503,15 @@ internal TfToken GuessInterpolation(int count)
             {
                 return UsdGeomTokens.constant;
             }
-            if (count == originalFaceVertexCounts.Length)
+            if (IsTopologyBackedUp() && count == originalFaceVertexCounts.Length)
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs:554]: Is there a scenario where GuessInterpolation /should/ be called, but the topology is NOT backed up?

If not, I think this should be an assert instead (and same a few lines down)

File: package/com.unity.formats.usd/Runtime/Scripts/IO/Scene/SceneImporter.cs [hunk 5]
```diff
@@ -813,6 +813,7 @@ private static void RemoveComponent<T>(GameObject go)
                                 GameObject go = primMap[pathAndSample.path];
                                 NativeImporter.ImportObject(scene, go, scene.GetPrimAtPath(pathAndSample.path),
                                     importOptions);
+                                // TODO: should we restore the conversion context here?
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/Scene/SceneImporter.cs]: Comment still valid?

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs [hunk 6]
```diff
@@ -87,15 +175,18 @@ public void Sanitize(Scene scene, SceneImportOptions importOptions)
             if (changeHandedness)
                 ConvertTransform();
 
-            var santizePrimvars = importOptions.ShouldBindMaterials ||
-                scene.IsPopulatingAccessMask || scene.AccessMask != null;                        //this is true when reading from the timeline
+            var santizePrimvars = IsRestored() && arePrimvarsFaceVarying || // if the sample is restored we already know
+                importOptions.ShouldBindMaterials ||
+                scene.IsPopulatingAccessMask ||                                  // this is true when initializign prims from the timeline
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs]: Comment typo

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs [hunk 7]
```diff
@@ -87,15 +175,18 @@ public void Sanitize(Scene scene, SceneImportOptions importOptions)
             if (changeHandedness)
                 ConvertTransform();
 
-            var santizePrimvars = importOptions.ShouldBindMaterials ||
-                scene.IsPopulatingAccessMask || scene.AccessMask != null;                        //this is true when reading from the timeline
+            var santizePrimvars = IsRestored() && arePrimvarsFaceVarying || // if the sample is restored we already know
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs]: I would add parens here to make this easier to read - for example, is it `IsRestore && arePrimvarsFaceVarying`, or is it the entire list of conditions?

Also, would be good to complete the sentence in the comment - already know what? That the primvars need to be sanitized?

Actually, is it possible to just remove IsRestored entirely from the API? I feel booleans like this tend to complicate state and lead to bugs...

File: package/com.unity.formats.usd/Runtime/Scripts/IO/ReadJob.cs [hunk 8]
```diff
@@ -96,7 +96,20 @@ public void Execute(int index)
...
-                sample.Sanitize(m_scene, m_importOptions);
+
+                var restorableSample = sample as IRestorable;
+                DeserializationContext deserializationContext = null;
+                m_scene.AccessMask?.Included.TryGetValue(m_paths[index], out deserializationContext);
+                if (restorableSample != null && deserializationContext != null)
+                {
+                    restorableSample.FromState(deserializationContext.state);
+                    sample.Sanitize(m_scene, m_importOptions);
+                    deserializationContext.state = restorableSample.ToState();
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/ReadJob.cs]: I would add a comment here, explaining how the lifetime works, something like:

```
// If a context was populated, it means streaming playback is currently happening.
// In addition, if the sample is of type IRestorable, then some state must be stored to support efficient streaming.
// This state is saved immediately after sanitizing the data and downstream processes are not allowed to modify this state.
```

File: package/com.unity.formats.usd/Dependencies/USD.NET/serialization/AccessMask.cs [hunk 9]
```diff
@@ -12,13 +12,43 @@
...
 namespace USD.NET
 {
+    /// <summary>
+    /// An interface to enable storage of the conversion state.
+    /// </summary>
+    /// <remarks>
+    /// In order to speed up the deserialization of animated prims we only load the dynamic properties. If the
+    /// downstream logic needs some static properties they can be stored temporarily in a ConversionContext.
+    /// </remarks>
+    public interface IConversionState
```
Comment (jcowles) [inline package/com.unity.formats.usd/Dependencies/USD.NET/serialization/AccessMask.cs]: Maybe `IPrimStreamingState`?

File: package/com.unity.formats.usd/Dependencies/USD.NET/serialization/AccessMask.cs [hunk 10]
```diff
@@ -28,6 +58,6 @@ namespace USD.NET
     /// </remarks>
     public class AccessMask
     {
-        public Dictionary<SdfPath, HashSet<MemberInfo>> Included = new Dictionary<SdfPath, HashSet<MemberInfo>>();
+        public Dictionary<SdfPath, DeserializationContext> Included = new Dictionary<SdfPath, DeserializationContext>();
```
Comment (jcowles) [inline package/com.unity.formats.usd/Dependencies/USD.NET/serialization/AccessMask.cs:58]: This field is named `Included`, does the serialization context belong here, or as a sibling to this field?

`AccessMask` is also no longer a great name for this - it should be something more like, `PrimStreamingContextMap`, or something.

We could make these changes in a future PR, but maybe add a comment?

File: package/com.unity.formats.usd/Dependencies/USD.NET/USD.NET.api [hunk 11]
```diff
@@ -14645,7 +14645,7 @@ namespace USD.NET
 {
     public class AccessMask
```
Comment (jcowles) [inline package/com.unity.formats.usd/Dependencies/USD.NET/USD.NET.api:14646]: Out of curiosity, are these .api files auto-generated?

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs [hunk 12]
```diff
@@ -9,6 +9,31 @@
 [assembly: InternalsVisibleToAttribute("Unity.Formats.USD.Tests")]
 namespace Unity.Formats.USD
 {
+    /// <summary>
+    /// When a prim is sparsely animated only dynamic properties are read into the Sample.
+    /// The ConversionSate purpose is to hold the data (static properties or computed data) to ensure proper conversion.
+    /// </summary>
+    public interface IRestorable
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs:17]: Do we need `IRestorable` and `IConversionState`?

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs [hunk 13]
```diff
@@ -17,6 +42,24 @@ public interface ISanitizable
         void Sanitize(Scene scene, SceneImportOptions importOptions);
     }
 
+    public class MeshConversionState : IConversionState
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs]: If you rename `IConversionState` to `IPrimStreamingState`, I think this should also be `MeshStreamingState`

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs [hunk 14]
```diff
@@ -9,6 +9,31 @@
...
+    /// <summary>
+    /// When a prim is sparsely animated only dynamic properties are read into the Sample.
+    /// The ConversionSate purpose is to hold the data (static properties or computed data) to ensure proper conversion.
+    /// </summary>
+    public interface IRestorable
+    {
+        /// <summary>
+        /// Returns true if the internal data has been restored from State data
+        /// </summary>
+        public bool IsRestored();
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs]: It would be awesome if we could remove this bool - for some reason, it triggers read flags for me, in terms of creating future bugs.

**Review by jcowles — APPROVED** [review]
I left a lot of comments, but they are mostly weak suggestions.

Overall, I think this is an awesome change - great work!

**Review by GeorgeIngUnity — APPROVED** [review]

File: package/com.unity.formats.usd/Dependencies/USD.NET/serialization/AccessMask.cs [hunk 15]
```diff
@@ -12,13 +12,43 @@
...
 using pxr;
 
 namespace USD.NET
 {
+    /// <summary>
+    /// An interface to enable storage of the conversion state.
+    /// </summary>
+    /// <remarks>
+    /// In order to speed up the deserialization of animated prims we only load the dynamic properties. If the
+    /// downstream logic needs some static properties they can be stored temporarily in a ConversionContext.
```
Comment (vickycl) [inline package/com.unity.formats.usd/Dependencies/USD.NET/serialization/AccessMask.cs]: ~ConversionContext no longer exists- should this be a DeserializationContext?~

File: package/com.unity.formats.usd/Dependencies/USD.NET/serialization/AccessMask.cs (hunk 9 above)
Comment (vickycl) [inline package/com.unity.formats.usd/Dependencies/USD.NET/serialization/AccessMask.cs]: I don't think calling this a state is very clear. A state implies it is the State of the Conversion, whereas it's more like a container for the data during the conversion. As we're storing the raw data deserialized from USD, maybe `IPrimDeserializedData`?

File: package/com.unity.formats.usd/Dependencies/USD.NET/serialization/AccessMask.cs [hunk 16]
```diff
@@ -12,13 +12,43 @@
...
+    {
+    }
+
+    /// <summary>
+    /// This class holds the data required by the deserializer to optimize reading of primitive with dynamic properties
+    /// </summary>
+    public class DeserializationContext : IEnumerable
+    {
+        public HashSet<MemberInfo> dynamicMembers;
+        public IConversionState state;
```
Comment (vickycl) [inline package/com.unity.formats.usd/Dependencies/USD.NET/serialization/AccessMask.cs]: Could we make this more clear that it stores the staticMembers, perhaps by calling it `public IPrimDeserializedData staticMemberData` or some such?

File: package/com.unity.formats.usd/Dependencies/USD.NET/serialization/AccessMask.cs (hunk 10 above)
Comment (vickycl) [reply package/com.unity.formats.usd/Dependencies/USD.NET/serialization/AccessMask.cs:58]: I agree AccessMask is vague and it would be good to have something clearer. I think Streaming implies something different though, unless I'm misunderstanding what is happening at import. Maybe `TimeVaryingDataMap`, something that calls out that it's specifically for data that is read over time. Then 'Included' makes a bit more sense as a name for the data that is time varying. Or even something more explicit- `TimeVaryingPrimToDeserializedDataMap`

File: package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs [hunk 17]
```diff
@@ -100,43 +118,43 @@ public void BeginReading(Scene scene, PrimMap primMap, SceneImportOptions import
...
+                            deserializationContext.dynamicMembers.Add(faceVertexIndices);
+                            deserializationContext.dynamicMembers.Add(orientation);
+                            deserializationContext.dynamicMembers.Add(points);
+                            deserializationContext.dynamicMembers.Add(normals);
+                            deserializationContext.dynamicMembers.Add(colors);
+                            deserializationContext.dynamicMembers.Add(st);
+                            deserializationContext.dynamicMembers.Add(uv);
+                            deserializationContext.dynamicMembers.Add(uv2);
+                            deserializationContext.dynamicMembers.Add(uv3);
+                            deserializationContext.dynamicMembers.Add(uv4);
```
Comment (vickycl) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs:129]: How come we need to add all of them if just one of the members is varying? Also could there be a better way of checking whether there's a varying one instead of searching the HashMap for each member?

I don't see anywhere where any of these types get added to the dynamicMembers outside of this function, so when will we hit this?

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs [hunk 18]
```diff
@@ -10,6 +10,31 @@
 [assembly: InternalsVisibleToAttribute("Unity.Formats.USD.Tests.Runtime")]
 namespace Unity.Formats.USD
 {
+    /// <summary>
+    /// When a prim is sparsely animated only dynamic properties are read into the Sample.
+    /// The ConversionSate purpose is to hold the data (static properties or computed data) to ensure proper conversion.
```
Comment (vickycl) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs]: Typo 'ConversionState'
Maybe expand in the comment specifically what `IRestorable` is for, ie copying static data into additional samples.

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs [hunk 19]
```diff
@@ -10,6 +10,31 @@
...
+    {
+        /// <summary>
+        /// Returns true if the internal data has been restored from State data
+        /// </summary>
+        bool IsRestored();
+
+        /// <summary>
+        /// Restore internal data from the State data
+        /// </summary>
+        void FromState(IConversionState state);
```
Comment (vickycl) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs]: If we rename the State rename these too

File: package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs [hunk 20]
```diff
@@ -457,7 +481,7 @@ public static void ImportSkinning(string path, SanitizedMeshSample usdMesh, Mesh
             //
             // Mesh Topology.
             //
-            if (options.meshOptions.topology == ImportMode.Import && usdMesh.faceVertexIndices != null)
```
Comment (vickycl) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs:460]: Should we check IsTopologyBackedUp() in here as we use originalFaceVertexCounts? I'm unsure whether this has introduced a route in which that could be null at this point?

File: package/com.unity.formats.usd/Runtime/Scripts/IO/ReadJob.cs [hunk 21]
```diff
@@ -96,7 +96,22 @@ public void Execute(int index)
             if (ShouldReadPath(m_scene, m_paths[index]))
             {
                 m_scene.Read(m_paths[index], sample);
-                sample.Sanitize(m_scene, m_importOptions);
+
+                var restorableSample = sample as IRestorable;
```
Comment (vickycl) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/ReadJob.cs:100]: I think we should check this isn't null straight away, as there's no point doing the 'TryGetValue' if so

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs [hunk 22]
```diff
@@ -10,6 +10,31 @@
 [assembly: InternalsVisibleToAttribute("Unity.Formats.USD.Tests.Runtime")]
 namespace Unity.Formats.USD
 {
+    /// <summary>
+    /// When a prim is sparsely animated only dynamic properties are read into the Sample.
+    /// The ConversionSate purpose is to hold the data (static properties or computed data) to ensure proper conversion.
+    /// </summary>
+    public interface IRestorable
```
Comment (vickycl) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs:17]: If we keep it, I think if this is the only usecase for it then 'restore' is a misleading name. It's more like caching the data for reuse across samples, so maybe could be renamed as such.

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs [hunk 23]
```diff
@@ -273,7 +376,8 @@ internal bool ShouldUnweldVertices(bool bindMaterials)
         {
             // If any primvar is face varying (1 value per vertex) or uniform (1 value per face), all  primvars + mesh attributes will have to be converted to face varying
             // TODO: expose interpolation for standard mesh attributes (normals, tangents)
-            return normals != null && (normals.Length == originalFaceVertexCounts.Length || normals.Length > points.Length) ||
+            return arePrimvarsFaceVarying ||
```
Comment (vickycl) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs]: This is getting unwieldly and hard to parse, can we break down the logic?

**Review by vickycl — COMMENTED** [review]
This is an awesome change but I found it hard to parse with the names chosen. I gave some suggestions that I think would make it easier to follow and therefore maintain :) 

General question- Are we generating extra Garbage with these changes?

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs (hunk 12 above)
Comment (mirceaispas) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs:17]: Because of how the assemblies are separated and used we need an extra layer of abstraction. Without this abstraction we will need to refactor how `USD.NET` and `Unity.Formats.USD` are interacting.

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs (hunk 7 above)
Comment (mirceaispas) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs]: The bool prevents us computing multiple times the same data.

File: package/com.unity.formats.usd/Dependencies/USD.NET/serialization/AccessMask.cs (hunk 15 above)
Comment (vickycl) [reply package/com.unity.formats.usd/Dependencies/USD.NET/serialization/AccessMask.cs]: After renaming I think this should be in an IRestorableData?

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs [hunk 24]
```diff
@@ -376,17 +376,16 @@ internal bool ShouldUnweldVertices(bool bindMaterials)
...
-                tangents != null &&
-                (tangents.Length == originalFaceVertexCounts.Length || tangents.Length > points.Length) ||
-                bindMaterials &&
-                (st.GetInterpolationToken() == UsdGeomTokens.faceVarying ||
-                    uv.GetInterpolationToken() == UsdGeomTokens.faceVarying ||
-                    uv2.GetInterpolationToken() == UsdGeomTokens.faceVarying ||
-                    uv3.GetInterpolationToken() == UsdGeomTokens.faceVarying ||
-                    uv4.GetInterpolationToken() == UsdGeomTokens.faceVarying);
+            bool shouldUnweldNormals = normals != null && (normals.Length == originalFaceVertexCounts.Length || normals.Length > points.Length);
+            bool shouldUweldColors = colors != null && (colors.GetInterpolationToken() == UsdGeomTokens.uniform || colors.GetInterpolationToken() == UsdGeomTokens.faceVarying);
```
Comment (michaeljblain) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs]: Unweld

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs [hunk 25]
```diff
@@ -273,16 +376,16 @@ internal bool ShouldUnweldVertices(bool bindMaterials)
...
```
Comment (vickycl) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs]: typo: Unweld :)

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs [hunk 26]
```diff
@@ -376,17 +376,16 @@ internal bool ShouldUnweldVertices(bool bindMaterials)
...
+            bool shouldUnweldTangents = tangents != null && (tangents.Length == originalFaceVertexCounts.Length || tangents.Length > points.Length);
+            bool isTextureMappingFaceVarying = st.GetInterpolationToken() == UsdGeomTokens.faceVarying ||
+                                               uv.GetInterpolationToken() == UsdGeomTokens.faceVarying ||
+                                               uv2.GetInterpolationToken() == UsdGeomTokens.faceVarying ||
+                                               uv3.GetInterpolationToken() == UsdGeomTokens.faceVarying ||
+                                               uv4.GetInterpolationToken() == UsdGeomTokens.faceVarying;
+
+            return arePrimvarsFaceVarying || shouldUnweldNormals || shouldUweldColors || shouldUnweldTangents || (bindMaterials && isTextureMappingFaceVarying);
```
Comment (michaeljblain) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs]: ShouldUnweldColors

File: package/com.unity.formats.usd/Tests/USD.NET/AccessMaskTests.cs [hunk 27]
```diff
@@ -48,22 +48,22 @@ public void ReadSample_WriteState_StateIsPreserved()
...
             scene.Read(primPath, newsample);
-            scene.AccessMask.Included[primPath].state = testState;
+            scene.AccessMask.Included[primPath].cachedData = testData;
 
             // Read a different frame and check the state
             scene.Time = 2;
             scene.IsPopulatingAccessMask = false;
             scene.Read(primPath, newsample);
-            var state = scene.AccessMask.Included[primPath].state;
+            var state = scene.AccessMask.Included[primPath].cachedData;
```
Comment (michaeljblain) [inline package/com.unity.formats.usd/Tests/USD.NET/AccessMaskTests.cs]: For naming consistency shouldn't state be named cachedData instead

**Comment by judubu** [comment]
This is great and LGTM. Thanks so much for taking it over the the finish line. 

Note that  https://github.com/Unity-Technologies/usd-unity-sdk/pull/310 into this PR (a single commit on top of the current branch) is somewhat related. This improves the loading of UVs by getting rid of the hardcoded uv attribute names and directly querying which attributes are used by the material assigned to a mesh.

**Review by vickycl — APPROVED** [review]
Happy to land this :)

**Review by michaeljblain — APPROVED** [review]

File: package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs (hunk 17 above)
Comment (vickycl) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs:129]: They are added implicitly, so this could be hit
